File: 01-OOP-Refresher/src/go/cat.go
package main

import "fmt"

type Cat struct {
	Name string
}

func (c *Cat) GetName() string { return c.Name }

func (c *Cat) Eat() {
	fmt.Printf("%s is eating.\n", c.Name)
}

func (c *Cat) MakeSound() {
	fmt.Printf("%s says Meow!\n", c.Name)
}

func (c *Cat) Purr() {
	fmt.Printf("%s is purring...\n", c.Name)
}

File: 01-OOP-Refresher/src/rust/cat.rs
use crate::animal::Animal;

pub struct Cat {
    name: String,
}

impl Cat {
    pub fn new(name: &str) -> Self {
        Cat {
            name: name.to_string(),
        }
    }

    pub fn purr(&self) {
        println!("{} is purring...", self.name);
    }
}

impl Animal for Cat {
    fn name(&self) -> &str {
        &self.name
    }

    fn make_sound(&self) {
        println!("{} says Meow!", self.name);
    }
}
